test/extended/util/disruption/disruption.go

```diff
@@ -280,9 +280,21 @@ func createTestFrameworks(tests []upgrades.Test) map[string]*framework.Framework
 // disruption flake if any disruption occurs, and uses reason to prefix the message. I.e. tolerate 0.1 of 10m total will fail
 // if the sum of the intervals is greater than 1m, or report a flake if any interval is found.
 func ExpectNoDisruption(f *framework.Framework, tolerate float64, total time.Duration, events monitor.EventIntervals, reason string) {
+	duration, describe := GetDisruption(events, "")
+	if percent := float64(duration) / float64(total); percent > tolerate {
+		framework.Failf("%s for at least %s of %s (%0.0f%%):\n\n%s", reason, duration.Truncate(time.Second), total.Truncate(time.Second), percent*100, strings.Join(describe, "\n"))
+	} else if duration > 0 {
+		FrameworkFlakef(f, "%s for at least %s of %s (%0.0f%%), this is currently sufficient to pass the test/job but not considered completely correct:\n\n%s", reason, duration.Truncate(time.Second), total.Truncate(time.Second), percent*100, strings.Join(describe, "\n"))
+	}
+}
+
+func GetDisruption(events monitor.EventIntervals, requiredLocator string) (time.Duration, []string) {
 	var duration time.Duration
 	var describe []string
 	for _, interval := range events {
+		if requiredLocator != "" && requiredLocator != interval.Locator {
+			continue
+		}
 		describe = append(describe, interval.String())
 		i := interval.To.Sub(interval.From)
```

Comment:
Can I buy some more letters?  I know it's prexisting, but you must know what this means.  What is `i` in this context?  An event summary of something?

This flow suggests some other entity is injecting failed connections into an event stream?

Reply:
Well, I thought you would understand the code more than I am :-).

Apparently, `Monitor` created here https://github.com/openshift/origin/blob/29b338da5bb7fc4309d5b4fe120e3cca61e1daae/pkg/monitor/api.go#L31
watches given Samplers (see `StartKubeAPIMonitoring()`) every second. Once the Sampler becomes ok -> failed -> back ok, Monitor records event "Kube API is not responding to GET requests" with:
* Locator: `kube-apiserver`
* From: time when the Monitor detected the Sampler went ok -> bad
* To: time when the Monitor detected the Sampler went bad -> ok

It took me a looong while to find out this. Still, you can actually see the events in the usual `Monitor cluster while tests execute` output:

```
Oct 12 15:24:26.863 E kube-apiserver Kube API started failing: namespaces "kube-system" not found
Oct 12 15:24:27.759 - 5s    E kube-apiserver Kube API is not responding to GET requests
Oct 12 15:24:33.861 I kube-apiserver Kube API started responding to GET requests
```

So the filter there catches "kube-apiserver Kube API is not responding to GET requests". BTW, you can see that the monitor realizes it almost 1 second after other part of the monitor detected the actual error. The whole outage was 7 seconds, not 5. The whole monitoring code is very confusing. There must have been some reason to make it so...